src/TArea.cpp

```diff
@@ -584,3 +606,372 @@ int TArea::createLabelId() const
+                    return;
+                }
+            }
+        }
+    }
+    if (currentRoomCount % 10 != 0) {
+        // Must add on any remainder otherwise the total will be wrong:
+        mpMap->incrementJsonProgressDialog(true, true, currentRoomCount % 10);
+    }
+    QJsonValue roomsValue{roomsArray};
+    areaObj.insert(scROOMS, roomsValue);
+    mpMap->getCurrentProgressRoomCount();
+
+    // Process the labels after the rooms so that the first area shows something
+    // quickly (from the rooms) even if it has a number of labels to do.
+
+    writeJsonLabels(areaObj);
+    QJsonValue areaValue{areaObj};
+    array.append(areaValue);
+}
+
+std::pair<int, QString> TArea::readJsonArea(const QJsonArray& array, const int areaIndex)
+{
+    const QJsonObject areaObj{array.at(areaIndex).toObject()};
+    const int id = areaObj.value(scID).toInt();
+    const QString name{areaObj.value(scNAME).toString()};
+    gridMode = areaObj.value(scGRID_MODE).toBool();
+    mUserData = readJsonUserData(areaObj.value(scUSER_DATA).toObject());
+    int roomCount = 0;
+    for (int roomIndex = 0, total = areaObj.value(scROOMS).toArray().count(); roomIndex < total; ++roomIndex) {
+        TRoom* pR = new TRoom(mpRoomDB);
+        int roomId = pR->readJsonRoom(areaObj.value(scROOMS).toArray(), roomIndex, id);
+        rooms.insert(roomId);
+        // This also sets the room id for the TRoom:
+        mpRoomDB->addRoom(roomId, pR, true);
+        if (++roomCount % 10 == 0) {
+            if (mpMap->incrementJsonProgressDialog(false, true, 10)) {
+                // Cancel has been hit - so give up straight away:
+                return {0, {}};
+            }
+        }
+    }
+    if (roomCount % 10 != 0) {
+        // Must add on any remainder otherwise the total will be wrong:
+        mpMap->incrementJsonProgressDialog(false, true, roomCount % 10);
+    }
+
+    if (areaObj.contains(scLABELS) && areaObj.value(scLABELS).isArray()) {
+        readJsonLabels(areaObj);
+    }
+    return {id, name};
+}
+
+void TArea::writeJsonUserData(QJsonObject& obj) const
+{
+    QJsonObject userDataObj;
+    if (mUserData.isEmpty()) {
+        // Skip creating a user data array if it will be empty:
+        return;
+    }
+    QMapIterator<QString, QString> itDataItem(mUserData);
+    while (itDataItem.hasNext()) {
+        itDataItem.next();
+        QJsonValue userDataValue{itDataItem.value()};
+        userDataObj.insert(itDataItem.key(), userDataValue);
+    }
+    const QJsonValue userDatasValue{userDataObj};
+    obj.insert(scUSER_DATA, userDatasValue);
+}
+
+// Takes a userData object and parses all its elements
+QMap<QString, QString> TArea::readJsonUserData(const QJsonObject& obj) const
+{
+    QMap<QString, QString> results;
+    if (obj.isEmpty()) {
+        // Skip doing anything more if there is nothing to do:
+        return results;
+    }
+
+    QStringList keys = obj.keys();
+    for (int i = 0, total = keys.count(); i < total; ++i) {
```

Comment:
Difficult to follow - can a simple ranged-for be used to read the object's key and value at once? This pattern is repeated elsewhere and should be simplified.